crates/dojo-client/src/sync.rs

```diff
@@ -0,0 +1,95 @@
+use std::sync::Arc;
+
+use starknet::core::types::{BlockId, BlockTag};
+use starknet::core::utils::cairo_short_string_to_felt;
+use starknet::providers::Provider;
+use starknet_crypto::FieldElement;
+use tokio::sync::RwLock;
+
+use crate::contract::component::ComponentReader;
+use crate::contract::world::WorldContractReader;
+use crate::storage::EntityStorage;
+
+pub struct EntityReader<'a, P: Provider + Sync> {
+    /// Component name
+    component: String,
+    /// The entity keys
+    entity_keys: Vec<FieldElement>,
+    /// Component reader
+    reader: ComponentReader<'a, P>,
+}
+
+/// Request to sync a component of an entity.
+/// This struct will be used to cosntruct an [EntityReader].
+pub struct EntityComponentRequest {
+    /// Component name
+    pub component: String,
+    /// The entity keys
+    pub keys: Vec<FieldElement>,
+}
+
+pub struct WorldPartialSyncer<'a, S: EntityStorage, P: Provider + Sync> {
+    /// We wrap it in an Arc<RwLock> to allow sharing the storage between threads.
+    storage: Arc<RwLock<S>>,
+    world_reader: &'a WorldContractReader<'a, P>,
+    /// The entity components to sync
+    entity_components_to_sync: Vec<EntityComponentRequest>,
+}
+
+impl<'a, S, P> WorldPartialSyncer<'a, S, P>
+where
+    S: EntityStorage + Send + Sync,
+    P: Provider + Sync + 'static,
+{
+    pub fn new(
+        storage: Arc<RwLock<S>>,
+        world_reader: &'a WorldContractReader<'a, P>,
+        entity_components_to_sync: Vec<EntityComponentRequest>,
+    ) -> WorldPartialSyncer<'a, S, P> {
+        Self { world_reader, storage, entity_components_to_sync }
+    }
+
+    // TODO: wrap this logic in a Future
+    /// start the syncing loop.
+    /// fetch the requested components values for every fixed interval.
+    pub async fn start(&mut self) {
+        let mut entity_readers = Vec::new();
+
+        for i in &self.entity_components_to_sync {
```

Comment:
In the final impl, we'll want a way to register new entities and components at runtime. For this simple example this is fine though